src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posix.c

```diff
@@ -175,18 +191,42 @@ static fd_node *pop_fd_node(fd_node **head, int fd) {
   return NULL;
 }
 
+/* Check if \a fd is still readable */
+static bool grpc_ares_is_fd_still_readable(grpc_ares_ev_driver *ev_driver,
+                                           int fd) {
+  ares_socket_t socks[ARES_GETSOCK_MAXNUM];
+  int socks_bitmask =
+      ares_getsock(ev_driver->channel, socks, ARES_GETSOCK_MAXNUM);
+  for (size_t i = 0; i < ARES_GETSOCK_MAXNUM; i++) {
+    if (ARES_GETSOCK_READABLE(socks_bitmask, i)) {
+      size_t bytes_available = 0;
+      return ioctl(fd, FIONREAD, &bytes_available) == 0 && bytes_available > 0;
+    }
+  }
+  return false;
+}
+
 static void on_readable_cb(grpc_exec_ctx *exec_ctx, void *arg,
                            grpc_error *error) {
   fd_node *fdn = (fd_node *)arg;
   grpc_ares_ev_driver *ev_driver = fdn->ev_driver;
   gpr_mu_lock(&fdn->mu);
```

Comment:
curious: isn't `fdn->ev_driver` protected by `fdn->mu`? Perhaps it's read only...

Reply:
`fdn->ev_driver` is not protected by `fdn->mu`. A `fd_node` can only be owned by one `ev_driver`. `fdn->ev_driver` is a read-only reference to this `ev_driver` object.